rosetta/docs/GPU_performance.md

```diff
@@ -0,0 +1,134 @@
+      
+- combining tensors that are sharded along different dimensions. Within a transformer layer, tensors can be sharded row-wise or column-wise and by default XLA will generate multiple collective calls for tensors sharded along different dimensions. The following optimization flags combine all tensors shardings, and map them to a group NCCL call that has a large commulative size and achieves high communication efficiency. 
+    - --xla_gpu_enable_all_gather_combine_by_dim=false
+    - --xla_gpu_enable_reduce_scatter_combine_by_dim=false
+      
+- Combine threshold values in XLA that determine when an all-gather (AG) or reduce-scatter (RS) is triggered. We want to set these values to be at least as large as the size of weights (AG) or gradients (RS) in a single transformer layer since large communication buffers achieve higher link bandwidth utilization. For example, LLAMA2-7B with BF16 weights and gradients, we have 32 transformer layers => each layer has ~218M weights => one would want to set these thresholds to at least 436MB.
+    - --xla_gpu_all_gather_combine_threshold_bytes=8589934592
+    - --xla_gpu_reduce_scatter_combine_threshold_bytes=8589934592
+      
+- Combine threshold values in XLA that determine when an all-reduce (AR) is triggered. Typically, used to overlap AR of gradients with back-prop of compute. We want to set this to be at least as large as possible to achieve high efficiency, but as small as possible to achieve maximum overlap. Depending on the interconnect of your system, one might want to try several threshold values in steps of 2 from say 16MB to total gradient size.
+    - --xla_gpu_all_reduce_combine_threshold_bytes=8589934592
+
+
+### Flags to enable async collective permute 
+
+The following flags enable overlap of pipeline parallel communication of send/recv with computation. 
+- --xla_gpu_enable_pipelined_p2p=true  (false by default)
+- --xla_gpu_collective_permute_decomposer_threshold=1024
+- --xla_gpu_lhs_enable_gpu_async_tracker=true
+
+### Flags to enable collective matmul
+
+The following flags enable overlap of tensor parallel communication with GEMMs/matmul by splicing GEMMs into smaller chunks and triggering each chunks' collective right after the chunk's GEMM is done. The threshold determines the size of output buffer of GEMM when this optimization becomes active (0 enables collective matmul for all GEMM-collective patterns)
+- --xla_gpu_multi_streamed_windowed_einsum=true
+- --xla_gpu_threshold_for_windowed_einsum_mib=0
+
+### Profile Guided Latency Estimator (PGLE)
+
+The following flag enables use of PGLE with JAX/XLA. Please see [PGLE notes](https://github.com/NVIDIA/JAX-Toolbox/blob/main/rosetta/docs/PGLE.md) for more details.
+- --xla_gpu_pgle_profile_file_or_directory_path=filename
+
+## Other XLA Flags 
+
+### Enable CUDA graphs
+
+The below enables CUDA Graph suppport for JAX/XLA workloads
+- --xla_gpu_enable_command_buffer (1 by default, available options are 0, 1, 2, and 3)
```

Comment:
Did you test that? From the code, that option take a string, not an int.
It was the legacy flag xla_gpu_graph_level that used int.

Reply:
Thanks, I made the change here. 

The below enables CUDA Graph suppport for JAX/XLA workloads, and is enabled by default.

--xla_gpu_enable_command_buffer
Valid options for operations that are supported by command buffer are FUSION, CUBLAS, COLLECTIVES, CONDITIONALS, CUSTOM_CALL. For example, --xla_gpu_enable_command_buffer=FUSION,CUBLAS,COLLECTIVES enables cuda graph for the respective operations.